optional simulate & wallet, editable TransactionBuilder

chadoh opened the pull request:
- Remove `TimeoutInfinite`, instead allowing people to set a `timeoutInSeconds` when they initialize an `AssembledTransaction`. Default: 10.

- No more `wallet` parameter. Instead, you pass:
   - `publicKey` – Will be used to construct the `Account` object that is used as the source of the transaction. Can be undefined, which means the null account will be used. Can be a string or, to ease operation with Freighter and other async-key-lookup in browser wallets, a promise of a string. 

   - `signTransaction`: matches the `signTransaction` signature from Freighter. Can be left blank when you initialize the ContractClient and only provided at the time that you call `signAndSend({ signTransaction: (...) => { ... } })`

   - `signAuthEntry`: matches the `signAuthEntry` signature from Freighter. Can also be left initially blank, but is needed if/when you call `signAuthEntries`.

 - You can now pass `simulate: false` when first creating your transaction to skip simulation. You can then modify the transaction using the TransactionBuilder at `tx.raw` before manually calling `simulate`. Example:

       const tx = await myContract.myMethod(
         { args: 'for', my: 'method', ... },
         { simulate: false }
       );
       tx.raw.addMemo(Memo.text('Nice memo, friend!'))
       await tx.simulate();

- Make `.simulation` getter public

- Error types are now collected under `AssembledTransaction.Errors` and
  `SentTransaction.Errors`.

- Move `ContractClient`, `AssembledTransaction` to separate entrypoints, so that they are not included in the global API or bundle size, but can still be packaged together with all the related stellar-sdk logic.

- Export `ExampleNodeWallet` from its own entrypoint, rather than hiding this logic entirely in the tests. @kalepail found this useful enough that he thought it would be worth exporting so other people can use it in their Node apps.

Shaptic commented on src/soroban/contract_client.ts:
```diff
@@ -42,14 +37,61 @@ export type ContractClientOptions = {
+ * If you want the transaction to be included in the ledger, you will need to
+ * provide a connected wallet.
+ */
+export async function getAccount(server: SorobanRpc.Server, wallet?: Wallet): Promise<Account> {
```
should/could this be the `AcceptsWalletOrAccount` interface, instead?

chadoh replied:
I'm not sure I understand the question. This function is used during initialization of the ContractClient and during initialization of an AssembledTransaction to get the account from the passed-in Wallet (or Signer, or whatever we want to call it), if you don't pass in an explicit `Account`. So it is used to support the `AcceptsWalletOrAccount` interface.

chadoh replied:
I got rid of `AcceptsWalletOrAccount` altogether now. See updated PR description.

Shaptic commented on src/soroban/contract_client.ts:
```diff
@@ -42,14 +37,61 @@ export type ContractClientOptions = {
-  ) {}
+  ) {
+    this.server = new SorobanRpc.Server(this.options.rpcUrl, {
+      allowHttp: this.options.rpcUrl.startsWith("http://"),
```
should we move on to forcing people to intentionally opt-in to insecure servers instead of doing it like this?

Shaptic commented on src/soroban/assembled_transaction.ts:
```diff
@@ -49,36 +54,78 @@ class Err<E extends ErrorMessage> implements Result<never, E> {
+  /**
+   * The maximum amount of time to wait for the transaction to complete. Default: {@link DEFAULT_TIMEOUT}
+   */
+  timeoutInSeconds?: number;
```
To move the discussion from here over:

> Do you think this is important enough to bring to people's attention that we should uglify every method call in this way, or do you think there's a good-enough default?

Honestly? Yes. I'm open to opinions from users but from a security standpoint, this should be an explicit choice every time. Also from a consistency standpoint, you have to call `setTimebounds`/`setTimeout` on your `TransactionBuilder` no matter what and intentionally opt-in to `TimeoutInfinite` if you decide to make that (bad) choice.

chadoh replied:
I want to make sure we're on the same page—this change removes the `TimeoutInfinite` default. People _would_ need to opt into that. The default is now ten seconds. 

If you already understood that, sorry for over-explaining. If that's new information, does it change your opinion?

Shaptic replied:
Ahh thanks for clarifying - I guess I missed that. I don't think there should be **any** default at all.

chadoh replied:
That's a huge change from current TS Bindings behavior and I'd like to hold off on it.

Shaptic replied:
works for me :+1: then a non-inf default is better

Shaptic commented on src/soroban/assembled_transaction.ts:
```diff
@@ -49,36 +54,78 @@ class Err<E extends ErrorMessage> implements Result<never, E> {
+  /**
+   * The result of the transaction simulation. This is set after the first call to `simulate`. It is difficult to serialize and deserialize, so it is not included in the `toJSON` and `fromJSON` methods. See `simulationResult` and `simulationTransactionData` for cached, serializable versions of the data needed by AssembledTransaction logic.
+   */
   private simulation?: SorobanRpc.Api.SimulateTransactionResponse;
```
pls wrap these comments to 80 or 100 cols :pray:

Shaptic commented on src/soroban/assembled_transaction.ts:
```diff
@@ -241,71 +304,55 @@ export class AssembledTransaction<T> {
 
     if (!force && this.isReadCall) {
-      throw new Error(
+      throw new AssembledTransaction.Errors.NoSignatureNeeded(
```
I don't get it - can't you just skip the sign step?

chadoh replied:
Yes, this is for if people call `signAndSend` on a read-only transaction. They have the option of forcing inclusion in the ledger with `force: true`

chadoh commented on test/e2e/src/util.js:
```diff
@@ -1,64 +1,89 @@
+    rpcUrl,
+    allowHttp: true,
+    publicKey: keypair.publicKey(),
+    ...wallet,
```
Note that the exposed interface no longer uses the word `wallet` at all, but people may still choose to name the thing which exports `signTransaction` and `signAuthEntry` a _wallet_.

chadoh commented:
@Shaptic I still don't have write access to this repository—could you merge if you approve? @kalepail you might also be able to merge?

kalepail commented:
I don't feel comfortable merging as I'm not reviewing code as much as functionality. I'll leave that for @Shaptic

Shaptic commented on src/contract_client.ts:
```diff
@@ -0,0 +1,118 @@
+   * account will be used for the transaction simulation. If the transaction
+   * needs to be signed and sent, this is required.
+   */
+  publicKey?: string | Promise<string>;
```
Can we get away with not making this optional? Like is there really any tangible benefit for a downstream user to be able to avoid passing a public key? They can just generate one if they are truly that lazy.

Shaptic commented on src/contract_client.ts:
```diff
@@ -0,0 +1,118 @@
+    opts?: {
+      accountToSign?: string;
+    }
+  ) => Promise<XDR_BASE64>;
```
These two are remarkably similar; what do you think about an even more generic

```typescript
function sign(value: XDR_BASE64): Promise<XDR_BASE64>;
```

Then these two become specific versions of that above one.

Shaptic replied:
Yep, great idea :+1:

Shaptic commented on src/contract_client.ts:
```diff
@@ -0,0 +1,118 @@
+    }
+  ) => Promise<XDR_BASE64>;
+  contractId: string;
+  networkPassphrase: string;
```
useful to make this type-check against `StellarBase.Networks`?

Shaptic commented on src/contract_client.ts:
```diff
@@ -0,0 +1,118 @@
+          args: spec.funcArgsToScVals(name, args),
+          ...options,
+          ...contractClient.options,
+          errorTypes: spec
```
a comment here would be :kissing: :pinched_fingers:

chadoh replied:
```js
  /**
   * This gets filled in automatically from the ContractSpec if you use
   * {@link ContractClient.generate} to create your ContractClient.
   *
   * Background: If the contract you're calling uses the `#[contracterror]`
   * macro to create an `Error` enum, then those errors get included in the
   * on-chain XDR that also describes your contract's methods. Each error will
   * have a specific number.
   *
   * A ContractClient makes method calls with an {@link AssembledTransaction}.
   * When one of these method calls encounters an error, `AssembledTransaction`
   * will first attempt to parse the error as an "official" `contracterror`
   * error, by using this passed-in `errorTypes` object. See
   * {@link AssembledTransaction#parseError}. If `errorTypes` is blank or no
   * matching error is found,, then it will throw the raw error.
   * @default {}
   */
```

Shaptic replied:
oh I totally just meant an inline code comment for whoever is the future dev to grok the reduce bit, but this is above and beyond. nice!

Shaptic commented on src/example_node_wallet.ts:
```diff
@@ -0,0 +1,27 @@
+import { Keypair, TransactionBuilder, hash } from '.'
```
can we put these all in a subfolder to group them together in the repo? I don't like having an example in the top-level

chadoh replied:
Great idea. Moved to the `contract_client` folder.

Shaptic commented on src/example_node_wallet.ts:
```diff
@@ -0,0 +1,27 @@
+ * applications. Feel free to use this as a starting point for your own
+ * Wallet/TransactionSigner implementation.
+ */
+export class ExampleNodeWallet {
```
suggestion: `BasicNodeSigner` to both remove the "example" implication and "wallet" nomenclature

Shaptic commented:
@chadoh lmk if you'd rather I merge this and you resolve feedback in a later PR! doesn't matter to me

chadoh commented:
@Shaptic ok, I addressed all your feedback here. I think this is ready to merge to `bindings`. I'll then open a PR from `bindings` back to `master`—I think it's probably good enough to get it merged! I really want to remove the stale logic that's hard-coded into TS Bindings and rely on this instead.

Shaptic reviewed (approved):
This is turning into a really strong set of tools! Thanks for bearing with all of the back-and-forth on the design :clap:

chadoh commented:
Thanks, @Shaptic! You'll have to merge, as I do not have permission.
